assets/js/aplus.js

```diff
@@ -500,28 +505,40 @@ $(function() {
         searchOptions: function(show_dropdown) {
             if (show_dropdown && this.result.is(":visible") === false) {
                 this.search.find("button").dropdown("toggle");
-                return;
             }
-            this.result.find("li:not(.not-found)").remove();
-            this.result.find("li.not-found").hide();
             var selector = "option";
             const query = this.field.val().trim();
+            const self = this;
             if (query.length > 0) {
-                selector += ":contains(" + this.field.val() + ")";
-            }
-            var opt = this.element.find(selector);
-            if (opt.size() === 0) {
-                this.result.find("li.not-found").show();
-            } else {
-                var self = this;
-                opt.slice(0, 20).each(function(index) {
-                    var li = $('<li><a data-value="' + $(this).attr("value") + '">' + $(this).text() + '</a></li>');
-                    li.find("a").on("click", function(event) {
-                        self.addSelection($(this).attr("data-value"), $(this).text());
-                    });
-                    self.result.append(li);
-                });
+              $.ajax({
+                url: this.api_url,
+                data: {
+                  "search": query,
+                },
+              }).done( function( data ) {
+                const numOfEntries = data.count;
+                self.result.empty();
+                if(numOfEntries > 0) {
+                  for (let i = 0; i < numOfEntries; i++) {
+                      const student_info = data.results[i];
+                      var li = $('<li></li>').click(function() {
+                        self.addSelection(student_info['id'], student_info['full_name'] + ", " + student_info['student_id']);
+                      });
+                      const text = $('<a></a>');
+                      text.text(student_info['full_name'] + ", " + student_info['student_id']);
```

Comment:
The new JS code works only with the User search, since the search results should have attributes like 'student_id' or 'full_name'. Otherwise, there should not be any problem to add the new AJAX search to other models as well. Currently the 'search-select' -widget is only used with models other than users in the `DeadlineRuleDeviationForm` with exercises and modules. The original efficiency issue was with downloading thousands of users in to the HTML template, so those fields don't necessarily need the AJAX search.

Reply:
The deviation forms and views are going to have somewhat major refactoring in the open PR #401.

Reply:
Let's change the search select AJAX JS so that it does not hardcode references to the User model. Basically, the attributes 'student_id' and 'full_name' can not be hardcoded in the JS since they refer to the User API output. Let's read those values from HTML data attributes so that only the server side needs to define which forms are for users and which forms are something else.